File: lib/qni/angle_expression.rb
# frozen_string_literal: true

module Qni
  # Internal representations for normalized angle expressions.
  module AngleTerm
    # Numeric literal angle.
    class NumericLiteral
      def initialize(value, text)
        @value = value
        @text = text
      end

      def radians(_variables = {})
        value
      end

      def to_s
        text
      end

      def concrete?
        true
      end

      private

      attr_reader :value, :text
    end

    # Variable-backed angle reference.
    class VariableReference
      def initialize(name, expression_class, error_class)
        @name = name
        @expression_class = expression_class
        @error_class = error_class
      end

      def radians(variables = {})
        resolved_value = variables.fetch(name) do
          raise error_class, "unresolved angle variable: #{name}"
        end
        angle = expression_class.new(resolved_value)
        raise error_class, "variable value must be concrete: #{name}" unless angle.concrete?

        angle.radians
      end

      def to_s
        name
      end

      def concrete?
        false
      end

      private

      attr_reader :name, :expression_class, :error_class
    end

    # Product of a scalar and another angle expression.
    class Product
      def initialize(coefficient, coefficient_text, inner)
        @coefficient = coefficient
        @coefficient_text = coefficient_text
        @inner = inner
      end

      def radians(variables = {})
        coefficient * inner.radians(variables)
      end

      def to_s
        "#{coefficient_text}*#{inner}"
      end

      def concrete?
        inner.concrete?
      end

      private

      attr_reader :coefficient, :coefficient_text, :inner
    end

    # Signed wrapper around another angle expression.
    class Signed
      def initialize(sign, sign_text, inner)
        @sign = sign
        @sign_text = sign_text
        @inner = inner
      end

      def radians(variables = {})
        sign * inner.radians(variables)
      end

      def to_s
        sign_text == '+' ? inner.to_s : "#{sign_text}#{inner}"
      end

      def concrete?
        inner.concrete?
      end

      private

      attr_reader :sign, :sign_text, :inner
    end
  end

  # Parses and normalizes angle expressions such as π/2, 3*pi/4, and 0.5.
  class AngleExpression
    # Raised when a gate angle cannot be parsed.
    class Error < StandardError; end

    # Represents a concrete multiple of π with optional sign and denominator.
    class PiTerm
      PATTERN = %r{
        \A
        (?<sign>[+-]?)
        (?:(?<coefficient>\d+(?:\.\d+)?)(?:\*)?)?
        (?:π|pi)
        (?:(?:/|_)(?<denominator>\d+(?:\.\d+)?))?
        \z
      }x

      def self.parse(value)
        match = PATTERN.match(value)
        return unless match

        new(match.named_captures.compact.transform_keys(&:to_sym))
      end

      def initialize(parts)
        @parts = parts
      end

      def radians(_variables = {})
        sign * coefficient * Math::PI / denominator
      end

      def to_s
        "#{sign_prefix}#{coefficient_prefix}π#{denominator_suffix}"
      end

      def concrete?
        true
      end

      private

      attr_reader :parts

      def sign
        parts.fetch(:sign) == '-' ? -1.0 : 1.0
      end

      def sign_prefix
        parts.fetch(:sign) == '-' ? '-' : ''
      end

      def coefficient
        coefficient_text.to_f
      end

      def coefficient_prefix
        coefficient_text == '1' ? '' : coefficient_text
      end

      def coefficient_text
        parts.fetch(:coefficient, '1')
      end

      def denominator
        denominator_text.to_f
      end

      def denominator_suffix
        return '' if denominator_text == '1'

        "/#{denominator_text}"
      end

      def denominator_text
        parts.fetch(:denominator, '1')
      end
    end

    IDENTIFIER_PATTERN = /\A[a-zA-Z_][a-zA-Z0-9_]*\z/
    NUMERIC_PATTERN = /\A[+-]?\d+(?:\.\d+)?\z/
    MULTIPLIED_PATTERN = /\A(?<coefficient>[+-]?\d+(?:\.\d+)?)\*(?<term>.+)\z/
    SIGNED_IDENTIFIER_PATTERN = /\A(?<sign>[+-])(?<identifier>[a-zA-Z_][a-zA-Z0-9_]*)\z/

    def self.numeric_expression(value)
      return unless value.match?(NUMERIC_PATTERN)

      AngleTerm::NumericLiteral.new(value.to_f, value)
    end

    def initialize(raw_value)
      @raw_value = raw_value
    end

    def radians(variables = {})
      parsed_expression_or_error.radians(variables)
    end

    def to_s
      parsed_expression_or_error.to_s
    end

    def concrete?
      parsed_expression&.concrete? || false
    end

    private

    attr_reader :raw_value

    def normalized
      @normalized ||= begin
        value = raw_value.to_s.delete(' ')
                         .gsub('θ', 'theta')
                         .gsub(/(?<=\d)theta/, '*theta')
        raise Error, 'angle is required' if value.empty?

        value
      end
    end

    def parsed_expression
      @parsed_expression ||= parsed_expression_for(normalized)
    end

    def parsed_expression_or_error
      parsed_expression || raise_invalid_angle
    end

    def parsed_expression_for(value)
      self.class.numeric_expression(value) ||
        signed_variable_expression(value) ||
        variable_expression(value) ||
        PiTerm.parse(value) ||
        product_expression(value)
    end

    def signed_variable_expression(value)
      match = SIGNED_IDENTIFIER_PATTERN.match(value)
      return unless match

      sign_text = match[:sign]
      AngleTerm::Signed.new(
        { '+' => 1.0, '-' => -1.0 }.fetch(sign_text),
        sign_text,
        AngleTerm::VariableReference.new(match[:identifier], self.class, Error)
      )
    end

    def variable_expression(value)
      return unless value.match?(IDENTIFIER_PATTERN)

      AngleTerm::VariableReference.new(value, self.class, Error)
    end

    def product_expression(value)
      match = MULTIPLIED_PATTERN.match(value)
      return unless match

      coefficient_text = match[:coefficient]
      inner_expression = self.class.new(match[:term])
      AngleTerm::Product.new(coefficient_text.to_f, coefficient_text, inner_expression)
    end

    def raise_invalid_angle
      raise Error, "invalid angle: #{normalized}"
    end
  end
end

File: src/angle_expression.ts
export class AngleExpressionError extends Error {}

const IDENTIFIER_PATTERN = /^[a-zA-Z_][a-zA-Z0-9_]*$/u;
const NUMERIC_PATTERN = /^[+-]?\d+(?:\.\d+)?$/u;
const MULTIPLIED_PATTERN = /^(?<coefficient>[+-]?\d+(?:\.\d+)?)\*(?<term>.+)$/u;
const PI_PATTERN =
  /^(?<sign>[+-]?)(?:(?<coefficient>\d+(?:\.\d+)?)(?:\*)?)?(?:π|pi)(?:(?:\/|_)(?<denominator>\d+(?:\.\d+)?))?$/u;
const SIGNED_IDENTIFIER_PATTERN = /^(?<sign>[+-])(?<identifier>[a-zA-Z_][a-zA-Z0-9_]*)$/u;

export function validAngleIdentifier(value: string): boolean {
  return IDENTIFIER_PATTERN.test(value);
}

interface AngleTerm {
  readonly concrete: boolean;
  readonly radians: (variables: Readonly<Record<string, string>>) => number;
  readonly text: string;
}

export class AngleExpression {
  private readonly rawValue: unknown;

  constructor(rawValue: unknown) {
    this.rawValue = rawValue;
  }

  concrete(): boolean {
    return this.parse().concrete;
  }

  radians(variables: Readonly<Record<string, string>> = {}): number {
    return this.parse().radians(variables);
  }

  toString(): string {
    return this.parse().text;
  }

  private normalized(): string {
    const value = String(this.rawValue)
      .replaceAll(' ', '')
      .replaceAll('θ', 'theta')
      .replace(/(?<=\d)theta/gu, '*theta');

    if (value.length === 0) {
      throw new AngleExpressionError('angle is required');
    }

    return value;
  }

  parse(): AngleTerm {
    const value = this.normalized();
    const term = parseAngleTerm(value);

    if (!term) {
      throw new AngleExpressionError(`invalid angle: ${value}`);
    }

    return term;
  }
}

function parseAngleTerm(value: string): AngleTerm | undefined {
  return (
    parseNumericLiteral(value) ??
    parseSignedIdentifier(value) ??
    parseVariableReference(value) ??
    parsePiTerm(value) ??
    parseProduct(value)
  );
}

function parseNumericLiteral(value: string): AngleTerm | undefined {
  if (!NUMERIC_PATTERN.test(value)) {
    return undefined;
  }

  return { concrete: true, radians: () => Number(value), text: value };
}

function parseSignedIdentifier(value: string): AngleTerm | undefined {
  const match = SIGNED_IDENTIFIER_PATTERN.exec(value);

  if (!match?.groups) {
    return undefined;
  }

  const groups = match.groups;
  const identifier = groups.identifier ?? '';
  return {
    concrete: false,
    radians: (variables) => variableRadians(identifier, variables) * (groups.sign === '-' ? -1 : 1),
    text: groups.sign === '-' ? `-${identifier}` : identifier
  };
}

function parseVariableReference(value: string): AngleTerm | undefined {
  if (!IDENTIFIER_PATTERN.test(value)) {
    return undefined;
  }

  return { concrete: false, radians: (variables) => variableRadians(value, variables), text: value };
}

function parsePiTerm(value: string): AngleTerm | undefined {
  const match = PI_PATTERN.exec(value);

  if (!match?.groups) {
    return undefined;
  }

  const sign = match.groups.sign === '-' ? '-' : '';
  const coefficient = match.groups.coefficient ?? '1';
  const coefficientPrefix = coefficient === '1' ? '' : coefficient;
  const denominator = match.groups.denominator ?? '1';
  const denominatorSuffix = denominator === '1' ? '' : `/${denominator}`;

  return {
    concrete: true,
    radians: () => signValue(sign) * Number(coefficient) * Math.PI / Number(denominator),
    text: `${sign}${coefficientPrefix}π${denominatorSuffix}`
  };
}

function parseProduct(value: string): AngleTerm | undefined {
  const match = MULTIPLIED_PATTERN.exec(value);

  if (!match?.groups) {
    return undefined;
  }

  const coefficient = match.groups.coefficient ?? '';
  const inner = new AngleExpression(match.groups.term).parse();

  return {
    concrete: inner.concrete,
    radians: (variables) => Number(coefficient) * inner.radians(variables),
    text: `${coefficient}*${inner.text}`
  };
}

function variableRadians(name: string, variables: Readonly<Record<string, string>>): number {
  const resolvedValue = variables[name];

  if (resolvedValue === undefined) {
    throw new AngleExpressionError(`unresolved angle variable: ${name}`);
  }

  const angle = new AngleExpression(resolvedValue);

  if (!angle.concrete()) {
    throw new AngleExpressionError(`variable value must be concrete: ${name}`);
  }

  return angle.radians();
}

function signValue(sign: string): number {
  return sign === '-' ? -1 : 1;
}
